sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala

```diff
@@ -899,4 +900,28 @@ class WholeStageCodegenSuite extends QueryTest with SharedSparkSession
       }
     }
   }
+
+  test("SPARK-47238: Test broadcast threshold for generated code") {
+    // case 1: threshold is -1, shouldn't broadcast since smaller than 0 means disabled.
+    // case 2: threshold is a larger number, shouldn't broadcast since not yet exceeded.
+    // case 3: threshold is 0, should broadcast since it's always smaller than generated code size.
+    Seq((-1, false), (1000000000, false), (0, true)).foreach { case (threshold, shouldBroadcast) =>
+      withSQLConf(SQLConf.WHOLESTAGE_BROADCAST_CLEANED_SOURCE_THRESHOLD.key -> threshold.toString) {
+        val df = Seq(0, 1, 2).toDF().groupBy("value").sum()
+        // Invoke tryBroadcastCleanedSource and make sure it returns the desired variables.
```

Comment:
how about we enable `conf.usePartitionEvaluator` and call `WholeStageCodegenExec.execute`, and cast the RDD to `MapPartitionsWithEvaluatorRDD`, and get and cast the `evaluatorFactory` to `WholeStageCodegenEvaluatorFactory`. Then we can check if it really uses broadcast. Otherwise the test is not very convincing as checking `tryBroadcastCleanedSource` does not prove we use broadcast at the end.

Reply:
Good idea, otherwise I was even thinking of logging something to validate the broadcast. However, due to the way `evaluatorFactory` and `cleanedSourceOpt` are declared, they are not directly accessible, and thus we would need to retrieve them using reflection (I prefer not to modify the visibility).